substrate/frame/scheduler/src/lib.rs

```diff
@@ -1157,24 +1181,30 @@ impl<T: Config> Pallet<T> {
 			return
 		}
 
-		let mut incomplete_since = now + One::one();
-		let mut when = IncompleteSince::<T>::take().unwrap_or(now);
```

Comment:
> With the `IncompleteSince` we iterate over blocks that might have no task to execute and this might make a situation with many incomplete blocks even worth. But probably not too much? One more read?

Yes, but then this becomes unbounded in case too many blocks are skipped. The idea behind using the `Queue` is to bound this to a sufficient number.

> How we determine the MaxScheduledBlocks bound?

This should be determined similar to the existing `MaxScheduledPerBlock`?

> Both solutions need a strategy for a situation when there are two many tasks that can not be completed and the task queue only grow. If such strategy not yet in place.

There is already a retry mechanism and the task is purged if the retry count is exceeded (even if failed).

Reply:
The `Queue` not only bounds how many blocks gonna be processed from the past. It bounds for how many blocks we can schedule. If the number is `50`, we can schedule only `50` jobs with distinct schedule time. 

The `MaxScheduledPerBlock` for me seems simpler to define. Because the block size its exiting constrain the system have. But how many distinct schedule time points you can have is something new. 

Retries work in case if a certain task fails while it's function call is being executed (not the scheduler fail). I meant a case when there are many (or few but too heavy) overdue tasks (task_block < now), so that the scheduler never (or needs too many time) to complete them and exist such overdue state to start processing tasks in time. Do we handle such case?

Reply:
> The Queue not only bounds how many blocks gonna be processed from the past. It bounds for how many blocks we can schedule. If the number is 50, we can schedule only 50 jobs with distinct schedule time

Indeed, I do not find it quite comfortable to run a `for` loop with `IncompleteSince` when there could be an unknown number of blocks passed between the successive runs. You could always keep the `MaxScheduledBlocks` on the higher side that would give you a similar experience?

> I meant a case when there are many (or few but too heavy) overdue tasks (task_block < now), so that the scheduler never (or needs too many time) to complete them and exist such overdue state to start processing tasks in time. Do we handle such case?

But this stays as an issue even in the current implementation? The change here just makes it bounded, so that the scheduling itself is blocked in such a case.

Reply:
Maybe we can put a quite big bound on the `MaxScheduledBlocks`, it is just a vec of block numbers.

Reply:
I agree, unbounded loop does not sound as a nice solution. My intention is to challenge here different solutions and find the best together. Not trying to claim that some of them is wrong. 

> Maybe we can put a quite big bound on the `MaxScheduledBlocks`, it is just a vec of block numbers.

@gui1117 @gupnik  what the big number you can think of? it also will affect the PoV. the number will be quite specific for the runtime, probably not smaller than `1000`. I also thing there are setup where it can be hard to make an envelope estimation for it. Extreme example, you open task scheduling for end users.

What if we set some parameter like `MaxIncompleteBlock`, for example if equal to `600`, all tasks that overdue for 1 hour just ignored?

Reply:
But still cannot come up with anything better than `Queue`. May be it can work in two different modes, one when we aer sure that we have the scheduler being executed on every block (e.g. Relay Chain), and another when we setup the scheduler on a parachain with RC block provider. If it wont complicated things too much. Additionally we can add the ~`MaxIncompleteBlock`~ `MaxStaleTaskAge` parameter.

Reply:
I see, indeed it is bad for PoV, as it is read every block.

The situation we want to fix is when the scheduler is using the relay chain block, and the parachain doesn't execute often.

(1) Maybe in this case the scheduler should use a different block provider with less granularity like `relay chain block / 100` so that when doing `IncompleteSince` it increments with a step of 100 relay chain block until it arrives to now.

(2) Or otherwise we can have a more complex structure for the queue. We cut the vector in chunck of 100 blocks.
So we have a double map with first key is `block number / 100` and second key is `block number % 100`, the value is a vector of length at most 100.

But still if the parachains wake up every month it can be not good. But at this point they should use (1).

Reply:
Added `MaxStaleTaskAge` as suggested. Thanks both.

Reply:
> I see, indeed it is bad for PoV, as it is read every block.

Also the task scheduling is affected. 

> The situation we want to fix is when the scheduler is using the relay chain block, and the parachain doesn't execute often.

I think we have more. Let's look to what we have/plan today. 

1. Relay Chain with scheduler and local block provider. No concerns. The new `Queue` is even redundant.
2. Parachain with scheduler as local block provider. No concerns. The new `Queue` is even redundant.

Reply:
> I see, indeed it is bad for PoV, as it is read every block.

Also the task scheduling is affected. 

> The situation we want to fix is when the scheduler is using the relay chain block, and the parachain doesn't execute often.

I think we have next cases today/planned soon:

1. Relay Chain with scheduler working with local block provider. No concerns. The new `Queue` is even redundant;
2. Parachain with scheduler working with local block provider. Same as (1);
3. Parachain with scheduler working with Relay Chain block provider;
3.1 runs scheduler on every second RC block, same as (1);
3.2 RC or Parachain for some reason is not producing blocks for 2 hours, we have `1200` blocks to iterate through.

We have a problem with (3.2) case only. On the current version (without `Queue`) it will eventually handle the overdue blocks (we can even calculate how many blocks it will take, lets say if there is no tasks scheduled in that period). With the `Queue` such situation as (3.2) gonna be handled well, but with a cost. 

I would look into numbers, if with the current version we can handle 2 hours of overdue in some reasonable time (lets say 10 blocks), then I think we are fine even with current solution, we just need tests for it. If not, may be we can introduce the `Queue` in a way that it can be disabled for (1) and (2) cases.

Reply:
I just checked that we currently use `Scheduler` only for the Governance related pallets. I think the related tasks should be better eventually processed than dropped if too old. So `MaxStaleTaskAge` should be at least optional.

Reply:
Yes the referenda pallet creates an alarm for every ref to check the voting turnout.

>We have a problem with (3.2) case only. On the current version (without Queue) it will eventually handle the overdue blocks (we can even calculate how many blocks it will take, lets say if there is no tasks scheduled in that period).

Depends on how many blocks are produced. I guess when we assume that the parachain will produce blocks at least as fast as it can advance the scheduler then yes.  
Playing devils advocate here since there could be parachains that only produce one block every two hours, which would get stuck without ever catching up the `IncompleteSince`.

Conceptually, I believe that a priority Queue is the right data structure. We try to evaluate an ordered list of tasks by their order. The issue with implementing this as a Vector is obviously the PoV.
  
Maybe we can implement the Queue as a B Tree? Then we can get the next task in log reads and insert in log writes. And it allows us to do exactly what we want: get the next pending task.  
To me it just seems that most of the pain here is that we are using the wrong data structure for the task.